internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2024 TiKV Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package unionstore
+
+import (
+	"sync"
+	"sync/atomic"
+
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/metrics"
+)
+
+// PipelinedMemDB is a Store which contains
+//   - a mutable buffer for read and write
+//   - an immutable onflushing buffer for read
+//   - like MemDB, PipelinedMemDB also CANNOT be used concurrently
+type PipelinedMemDB struct {
+	// Like MemDB, this RWMutex only used to ensure memdbSnapGetter.Get will not race with
+	// concurrent memdb.Set, memdb.SetWithFlags, memdb.Delete and memdb.UpdateFlags.
+	sync.RWMutex
+	onFlushing        atomic.Bool
+	errCh             chan error
+	flushFunc         FlushFunc
+	memBuffer         *MemDB
+	flushingMemBuffer *MemDB
+	len, size         int // len and size records the total flushed and onflushing memdb.
+}
+
+const (
+	MinFlushKeys = 10000
+	MinFlushSize = 16 * 1024 * 1024  // 16MB
+	MaxFlushSize = 128 * 1024 * 1024 // 128MB
+)
+
+type FlushFunc func(*MemDB) error
+
+func NewPipelinedMemDB(flushFunc FlushFunc) *PipelinedMemDB {
+	return &PipelinedMemDB{
+		memBuffer: newMemDB(),
+		errCh:     make(chan error, 1),
+		flushFunc: flushFunc,
+	}
+}
+
+// Dirty returns whether the pipelined buffer is mutated.
+func (p *PipelinedMemDB) Dirty() bool {
+	return p.memBuffer.Dirty() || p.flushingMemBuffer != nil
+}
+
+// Get the value by given key, it returns tikverr.ErrNotExist if not exist.
+// The priority of the value is MemBuffer > flushingMemBuffer MemBuffer > Snapshot.
+func (p *PipelinedMemDB) Get(k []byte) ([]byte, error) {
+	v, err := p.memBuffer.Get(k)
+	if p.flushingMemBuffer != nil && tikverr.IsErrNotFound(err) {
+		v, err = p.flushingMemBuffer.Get(k)
+	}
+	if err != nil {
+		return v, err
+	}
+	if len(v) == 0 {
+		return nil, tikverr.ErrNotExist
+	}
```

Comment:
Because `PipelinedMemDB` implement `Getter` interface, it doesn't accept a go context, it's more like a fast in-mem reader, I currently implement buffer_batch_get in upper layer.

Maybe we need to change the function to `func (p *PipelinedMemDB) Get(context.Context, []byte, tier int) ([]byte, error) {...}`?

Reply:
Add the buffer_batch_get, and we can skip it when lazy check existence when TiKV handling flush requests.